Crash in config.(*BuilderGCFilter).UnmarshalJSON

### Description

We used the Fuzz engine to modify some Test(TestBuilderGC) data, and then the following crash appeared. We hope to get the help of the developer to confirm whether it is a real bug.

### Reproduce

Part of the modified Test Code

```go
	tempFile := fs.NewFile(t, "config", fs.WithContent(`{
  "builder": {
    "gc": {
      "enabled": true,
      "poliCy": [
        {"keepStorage": "10GB", "filter": ["unused-for(2200h"]},
        {"keepStorage": "50GB", "filter": {"unused-for": {"3300h": true}}},
        {"keepStorage": "100GB", "all": true}
      ]
    }
  }
}`))
	defer tempFile.Remove()
	configFile := tempFile.Path()

	cfg, err := MergeDaemonConfigurations(&Config{}, nil, configFile) //crash in MergeDaemonConfigurations
```

Crash Log

```
panic: runtime error: index out of range [1] with length 1 [recovered]
	panic: runtime error: index out of range [1] with length 1

goroutine 101 [running]:
testing.tRunner.func1.2({0x11f1120, 0xc000294cc0})
	/home/gogen/.local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/home/gogen/.local/go/src/testing/testing.go:1212 +0x218
panic({0x11f1120, 0xc000294cc0})
	/home/gogen/.local/go/src/runtime/panic.go:1038 +0x215
github.com/docker/docker/daemon/config.(*BuilderGCFilter).UnmarshalJSON(0xc000298108, {0xc0002a3870, 0x14, 0x190})
	/home/gogen/workspace/gowork/src/topproj/moby/daemon/config/builder.go:50 +0x338
encoding/json.(*decodeState).array(0xc0002fe090, {0x1175160, 0xc000298108, 0x1a2})
	/home/gogen/.local/go/src/encoding/json/decode.go:506 +0x3f6
encoding/json.(*decodeState).value(0xc0002fe090, {0x1175160, 0xc000298108, 0xc0002a3867})
	/home/gogen/.local/go/src/encoding/json/decode.go:363 +0x7e
encoding/json.(*decodeState).object(0xc0002fe090, {0x11b3a40, 0xc000298100, 0xb})
	/home/gogen/.local/go/src/encoding/json/decode.go:774 +0xca5
encoding/json.(*decodeState).value(0xc0002fe090, {0x11b3a40, 0xc000298100, 0x1})
	/home/gogen/.local/go/src/encoding/json/decode.go:373 +0x45
encoding/json.(*decodeState).array(0xc0002fe090, {0x10f5700, 0xc000b58960, 0x1d2})
	/home/gogen/.local/go/src/encoding/json/decode.go:561 +0x626
encoding/json.(*decodeState).value(0xc0002fe090, {0x10f5700, 0xc000b58960, 0xc0002a383b})
	/home/gogen/.local/go/src/encoding/json/decode.go:363 +0x7e
encoding/json.(*decodeState).object(0xc0002fe090, {0x11b3980, 0xc000b58958, 0x1192f01})
	/home/gogen/.local/go/src/encoding/json/decode.go:774 +0xca5
encoding/json.(*decodeState).value(0xc0002fe090, {0x11b3980, 0xc000b58958, 0xc0002a3816})
	/home/gogen/.local/go/src/encoding/json/decode.go:373 +0x45
encoding/json.(*decodeState).object(0xc0002fe090, {0x1192f60, 0xc000b58958, 0x1a6af01})
	/home/gogen/.local/go/src/encoding/json/decode.go:774 +0xca5
encoding/json.(*decodeState).value(0xc0002fe090, {0x1192f60, 0xc000b58958, 0x7})
	/home/gogen/.local/go/src/encoding/json/decode.go:373 +0x45
encoding/json.(*decodeState).object(0xc0002fe090, {0x120bfc0, 0xc000b58580, 0x886305})
	/home/gogen/.local/go/src/encoding/json/decode.go:774 +0xca5
encoding/json.(*decodeState).value(0xc0002fe090, {0x120bfc0, 0xc000b58580, 0x90})
	/home/gogen/.local/go/src/encoding/json/decode.go:373 +0x45
encoding/json.(*decodeState).unmarshal(0xc0002fe090, {0x120bfc0, 0xc000b58580})
	/home/gogen/.local/go/src/encoding/json/decode.go:180 +0x1de
encoding/json.Unmarshal({0xc0002a3800, 0x7f609b0b0c00, 0x857d50}, {0x120bfc0, 0xc000b58580})
	/home/gogen/.local/go/src/encoding/json/decode.go:107 +0x125
github.com/docker/docker/daemon/config.getConflictFreeConfiguration({0xc0002942d0, 0xc000b58000}, 0x0)
	/home/gogen/workspace/gowork/src/topproj/moby/daemon/config/config.go:473 +0x405
github.com/docker/docker/daemon/config.MergeDaemonConfigurations(0x1390f08, 0xc000882820, {0xc0002942d0, 0x6})
	/home/gogen/workspace/gowork/src/topproj/moby/daemon/config/config.go:393 +0x33
```



### Expected behavior

-

### docker version

```bash
-
```


### docker info

```bash
-
```


### Additional Info

Implemented only with go test

--- a/daemon/config/builder.go
+++ b/daemon/config/builder.go
@@ -2,7 +2,6 @@ package config
 
 import (
 	"encoding/json"
-	"fmt"
 	"sort"
 	"strings"
 
@@ -28,7 +27,7 @@ func (x *BuilderGCFilter) MarshalJSON() ([]byte, error) {
 	for _, k := range keys {
 		values := f.Get(k)
 		for _, v := range values {
-			arr = append(arr, fmt.Sprintf("%s=%s", k, v))
+			arr = append(arr, k+"="+v)
 		}
 	}
 	return json.Marshal(arr)
@@ -45,9 +44,9 @@ func (x *BuilderGCFilter) UnmarshalJSON(data []byte) error {
 		return err
 	}
 	for _, s := range arr {
-		fields := strings.SplitN(s, "=", 2)
-		name := strings.ToLower(strings.TrimSpace(fields[0]))
-		value := strings.TrimSpace(fields[1])
+		name, value, _ := strings.Cut(s, "=")
+		name = strings.ToLower(strings.TrimSpace(name))
+		value = strings.TrimSpace(value)
 		f.Add(name, value)
 	}
 	*x = BuilderGCFilter(f)

--- a/daemon/config/builder_test.go
+++ b/daemon/config/builder_test.go
@@ -1,6 +1,7 @@
 package config
 
 import (
+	"encoding/json"
 	"testing"
 
 	"github.com/docker/docker/api/types/filters"
@@ -41,4 +42,17 @@ func TestBuilderGC(t *testing.T) {
 	// double check to please the skeptics
 	assert.Assert(t, filters.Args(cfg.Builder.GC.Policy[0].Filter).UniqueExactMatch("unused-for", "2200h"))
 	assert.Assert(t, filters.Args(cfg.Builder.GC.Policy[1].Filter).UniqueExactMatch("unused-for", "3300h"))
+}
+
+// TestBuilderGCFilterUnmarshal is a regression test for https://github.com/moby/moby/issues/44361,
+// where and incorrectly formatted gc filter option ("unused-for2200h",
+// missing a "=" separator). resulted in a panic during unmarshal.
+func TestBuilderGCFilterUnmarshal(t *testing.T) {
+	var cfg BuilderGCConfig
+	err := json.Unmarshal([]byte(`{"poliCy": [{"keepStorage": "10GB", "filter": ["unused-for2200h"]}]}`), &cfg)
+	assert.Check(t, err)
+	expectedPolicy := []BuilderGCRule{{
+		KeepStorage: "10GB", Filter: BuilderGCFilter(filters.NewArgs(filters.Arg("unused-for2200h", ""))),
+	}}
+	assert.DeepEqual(t, cfg.Policy, expectedPolicy, cmp.AllowUnexported(BuilderGCFilter{}))
 }

--- a/daemon/config/config.go
+++ b/daemon/config/config.go
@@ -313,13 +313,13 @@ func New() (*Config, error) {
 func GetConflictFreeLabels(labels []string) ([]string, error) {
 	labelMap := map[string]string{}
 	for _, label := range labels {
-		stringSlice := strings.SplitN(label, "=", 2)
-		if len(stringSlice) > 1 {
+		key, val, ok := strings.Cut(label, "=")
+		if ok {
 			// If there is a conflict we will return an error
-			if v, ok := labelMap[stringSlice[0]]; ok && v != stringSlice[1] {
-				return nil, errors.Errorf("conflict labels for %s=%s and %s=%s", stringSlice[0], stringSlice[1], stringSlice[0], v)
+			if v, ok := labelMap[key]; ok && v != val {
+				return nil, errors.Errorf("conflict labels for %s=%s and %s=%s", key, val, key, v)
 			}
-			labelMap[stringSlice[0]] = stringSlice[1]
+			labelMap[key] = val
 		}
 	}
 

